plasTeX/ConfigManager.py

```diff
@@ -0,0 +1,181 @@
+from argparse import _ArgumentGroup as ArgumentGroup, ArgumentParser
+from abc import ABC, abstractmethod
+from typing import Dict, List, Generic, TypeVar, Any, Optional, Type, KeysView, Union
+from configparser import ConfigParser
+import shlex
+
+T = TypeVar('T')
+
+def formatDefault(s: Any):
+    return " [" + str(s).replace("%", "%%") + "]"
+
+class Option(Generic[T]):
+    def __init__(self, description: str, options: str, default: T):
+        self.description = description
+        self.options = options.split(" ")
+        self.name = self.options[0].lstrip("-")
+        self._value = default
+
+    def registerArgparse(self, group: ArgumentGroup):
+        group.add_argument(*self.options, dest=self.name, type=self.valueType(), help=self.description + formatDefault(self.value))
+
+    def valueType(self) -> Type:
+        return type(self.value)
+
+    def updateFromDict(self, data: Dict["str", Any]):
+        value = data.get(self.name)
+        if value is not None:
+            self.value = value
+
+    def setFromString(self, string: str):
+        self.value = self.valueType()(string)
+
+    @property
+    def value(self) -> T:
+        return self._value
+
+    @value.setter
+    def value(self, value: T):
+        if type(value) != self.valueType():
+            raise TypeError("Encountered {} when expecting {}".format(type(value).__name__, self.valueType().__name__))
+
+        self._value = value
+
+class BooleanOption(Option[bool]):
+    def registerArgparse(self, group: ArgumentGroup):
+        enables = [x for x in self.options if x[0] != "!"]
+        disables = [x[1:] for x in self.options if x[0] == "!"]
+
+        group.add_argument(*enables, dest=self.name, help=self.description + formatDefault(self.value), action='store_true', default=None)
+        if len(disables) > 0:
+            group.add_argument(*disables, dest=self.name, help=self.description + formatDefault(self.value), action='store_false', default=None)
+
+class MultiStringOption(Option[List[str]]):
+    def registerArgparse(self, group: ArgumentGroup):
+        group.add_argument(*self.options, dest=self.name, type=str, nargs="*", help=self.description, action="append")
+
+    def setFromString(self, string: str):
+        self.value.extend(shlex.split(string))
+
+    def updateFromDict(self, data: Dict["str", Any]):
+        value = data.get(self.name)
+        if value is not None:
```

Comment:
You could use `value = data.get(self.name, [])` and skip the if. I think the intent would be clearer.

Reply:
Done.

Reply:
Okay so this actually doesn't work, because `data[self.name]` could take the value `None`.